apollo-client/src/main/java/com/ctrip/framework/apollo/monitor/internal/listener/impl/DefaultApolloClientBootstrapArgsApi.java

```diff
@@ -46,50 +49,52 @@ public class DefaultApolloClientBootstrapArgsApi extends
+    putAttachmentValue(APOLLO_CACHE_DIR, configUtil.getDefaultLocalCacheDir());
+    putAttachmentValue(APOLLO_CLUSTER, configUtil.getCluster());
+    putAttachmentValue(APOLLO_CONFIG_SERVICE,
         System.getProperty(APOLLO_CONFIG_SERVICE));
-    bootstrapArgs.put(APOLLO_CLIENT_MONITOR_EXTERNAL_TYPE, configUtil.getMonitorExternalType());
-    bootstrapArgs.put(APOLLO_CLIENT_MONITOR_ENABLED, configUtil.isClientMonitorEnabled());
-    bootstrapArgs.put(APOLLO_CLIENT_MONITOR_EXTERNAL_EXPORT_PERIOD,
+    putAttachmentValue(APOLLO_CLIENT_MONITOR_EXTERNAL_TYPE, configUtil.getMonitorExternalType());
+    putAttachmentValue(APOLLO_CLIENT_MONITOR_ENABLED, configUtil.isClientMonitorEnabled());
+    putAttachmentValue(APOLLO_CLIENT_MONITOR_EXTERNAL_EXPORT_PERIOD,
         configUtil.getMonitorExternalExportPeriod());
-    bootstrapArgs.put(APOLLO_META, configUtil.getMetaServerDomainName());
-    bootstrapArgs.put(APOLLO_PROPERTY_NAMES_CACHE_ENABLE, configUtil.isPropertyNamesCacheEnabled());
-    bootstrapArgs.put(APOLLO_PROPERTY_ORDER_ENABLE, configUtil.isPropertiesOrderEnabled());
-    bootstrapArgs.put(APOLLO_CLIENT_MONITOR_JMX_ENABLED, configUtil.isClientMonitorJmxEnabled());
-    bootstrapArgs.put(APOLLO_CLIENT_MONITOR_EXCEPTION_QUEUE_SIZE,
+    putAttachmentValue(APOLLO_META, configUtil.getMetaServerDomainName());
+    putAttachmentValue(APOLLO_PROPERTY_NAMES_CACHE_ENABLE, configUtil.isPropertyNamesCacheEnabled());
+    putAttachmentValue(APOLLO_PROPERTY_ORDER_ENABLE, configUtil.isPropertiesOrderEnabled());
+    putAttachmentValue(APOLLO_CLIENT_MONITOR_JMX_ENABLED, configUtil.isClientMonitorJmxEnabled());
+    putAttachmentValue(APOLLO_CLIENT_MONITOR_EXCEPTION_QUEUE_SIZE,
         configUtil.getMonitorExceptionQueueSize());
-    bootstrapArgs.put(APP_ID, configUtil.getAppId());
-    bootstrapArgs.put(ENV, configUtil.getApolloEnv());
-    bootstrapArgs.put(VERSION, Apollo.VERSION);
-    bootstrapArgs.forEach((key, value) -> {
-      if (value != null) {
-        bootstrapArgsString.put(key, value.toString());
-      }
-    });
+    putAttachmentValue(APP_ID, configUtil.getAppId());
+    putAttachmentValue(ENV, configUtil.getApolloEnv());
+    putAttachmentValue(VERSION, Apollo.VERSION);
+    putAttachmentValue(META_FRESH, DateUtil.formatLocalDateTime(LocalDateTime.now()));
+    putAttachmentValue(CONFIG_SERVICE_URL,"");
 
   }
 
   @Override
   public void collect0(ApolloClientMonitorEvent event) {
     String argName = event.getName();
     if (bootstrapArgs.containsKey(argName)) {
-      bootstrapArgs.put(argName, event.getAttachmentValue(argName));
+      putAttachmentValue(argName, event.getAttachmentValue(argName));
     } else {
       logger.warn("Unhandled event name: {}", argName);
     }
   }
+  
+  private void putAttachmentValue(String argName, Object value) {
+    bootstrapArgs.put(argName, value);
+    bootstrapArgsString.put(argName, value == null ? null : value.toString());
```

Comment:
Maybe we need to put only and only if value is not null

if value is null, ignore it is a better choice?

because null will cause more problem when get the value

Reply:
ac